internal/promotion/runner/builtin/yaml_merger.go

```diff
@@ -0,0 +1,134 @@
+package builtin
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	securejoin "github.com/cyphar/filepath-securejoin"
+	"github.com/xeipuuv/gojsonschema"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/yaml"
+	"github.com/akuity/kargo/pkg/promotion"
+	"github.com/akuity/kargo/pkg/x/promotion/runner/builtin"
+)
+
+// yamlMerger is an implementation of the promotion.StepRunner interface that
+// merges multiple YAML files.
+type yamlMerger struct {
+	schemaLoader gojsonschema.JSONLoader
+}
+
+// newYAMLMerger returns an implementation of the promotion.StepRunner interface
+// that merges multiple YAML files.
+func newYAMLMerger() promotion.StepRunner {
+	r := &yamlMerger{}
+	r.schemaLoader = getConfigSchemaLoader(r.Name())
+	return r
+}
+
+// Name implements the promotion.StepRunner interface.
+func (y *yamlMerger) Name() string {
+	return "yaml-merge"
+}
+
+// Run implements the promotion.StepRunner interface.
+func (y *yamlMerger) Run(
+	ctx context.Context,
+	stepCtx *promotion.StepContext,
+) (promotion.StepResult, error) {
+	cfg, err := y.convert(stepCtx.Config)
+	if err != nil {
+		return promotion.StepResult{
+			Status: kargoapi.PromotionStepStatusFailed,
+		}, &promotion.TerminalError{Err: err}
+	}
+	return y.run(ctx, stepCtx, cfg)
+}
+
+// convert validates yamlMerger configuration against a JSON schema and
+// converts it into a builtin.YAMLMergeConfig struct.
+func (y *yamlMerger) convert(cfg promotion.Config) (builtin.YAMLMergeConfig, error) {
+	return validateAndConvert[builtin.YAMLMergeConfig](y.schemaLoader, cfg, y.Name())
+}
+
+// validate validates yamlMerger configuration against a JSON schema.
+func (y *yamlMerger) validate(cfg promotion.Config) error {
+	return validate(y.schemaLoader, gojsonschema.NewGoLoader(cfg), y.Name())
+}
+
+func (y *yamlMerger) run(
+	_ context.Context,
+	stepCtx *promotion.StepContext,
+	cfg builtin.YAMLMergeConfig,
+) (promotion.StepResult, error) {
+
+	// sanity check
+	if len(cfg.InFiles) == 0 || cfg.OutFile == "" {
+		return promotion.StepResult{Status: kargoapi.PromotionStepStatusErrored}, fmt.Errorf("inFiles and OutFile must not be empty")
+	}
```

Comment:
Wouldn't validation against the schema already have caught either of these issues?